Mailing feature

JiahuiChen99 opened the pull request:
### **New Feature**
Each generated certificate gets sent to the recipient. It uses Yagmail, for now only Gmail senders are supported.
Dependencies updated to the latest version, almost all of them.
> PDF form fields **MUST** have the "Read Only - Sólo Lectura" option checked!

### **Update**
Replaced `PDFJinja` with `pdfrw`.
Fixed a variable comparison bug
```
len_fields2fill = len(fields2fill)
        len_csv_columns = len(next(csv_data))
        if len_fields2fill is not len_csv_columns:
            return "Columns error"
```

JiahuiChen99 committed:
chore: updated dependencies and imports to the latest version

JiahuiChen99 committed:
chore: endpoint render refactor

JiahuiChen99 committed:
chore: serve staticfiles renamed to static

JiahuiChen99 committed:
chore: django new path pattern

JiahuiChen99 committed:
chore: pdfrw dependency

JiahuiChen99 committed:
feat: creates a PDF Certificate for each participant

JiahuiChen99 committed:
chore: add pdfrw import

JiahuiChen99 committed:
fix: overriding annotations bug

JiahuiChen99 committed:
feat: Fill each annotation with the CSV corresponding data

JiahuiChen99 committed:
feat: make the annotation non-editable for the user

JiahuiChen99 committed:
feat: UTF-8 support

JiahuiChen99 committed:
Merge pull request #1 from JiahuiChen99/fix/dependencies

Solved dependencies & PDF generator

JiahuiChen99 committed:
chore: mailer & env variables dependencies

JiahuiChen99 committed:
feat: mailer instance & env variables loading

JiahuiChen99 committed:
chore: ignoring environment variables

JiahuiChen99 committed:
feat: mail data
- Destination
- Body
- Subject

JiahuiChen99 committed:
style: some refactoring and doc

JiahuiChen99 committed:
feat: send mail with data

JiahuiChen99 committed:
feat: ignore some UTF-8 parsing errors

JiahuiChen99 committed:
feat: changed regex for file compression
- yyy_teem.pdf

JiahuiChen99 committed:
Merge pull request #2 from JiahuiChen99/feat/mailing

Feat/mailing

JiahuiChen99 committed:
Merge pull request #3 from JiahuiChen99/dev

Mailing feature

JiahuiChen99 committed:
chore: lock form field commented
- PDF form field MUST be "Read Only"

JiahuiChen99 committed:
Merge pull request #4 from JiahuiChen99/hotfix/block_form

chore: lock form field commented

cbjuan commented on django_pdf_form_filler/settings.py:
```diff
@@ -45,15 +45,16 @@ def relative_project_path(*x):
     'django.contrib.messages.middleware.MessageMiddleware',
     'django.middleware.clickjacking.XFrameOptionsMiddleware',
     'django.middleware.security.SecurityMiddleware',
+    # 'whitenoise.middleware.WhiteNoiseMiddleware',
```
This may be removed
```suggestion
```

cbjuan commented on django_pdf_form_filler/settings.py:
```diff
@@ -74,7 +75,7 @@ def relative_project_path(*x):
 ]
 
-WSGI_APPLICATION = 'django_pdf_form_filler.wsgi.application'
+# WSGI_APPLICATION = 'application'
```
If not used, it may be removed
```suggestion
```

cbjuan commented on generator/views.py:
```diff
@@ -78,36 +82,56 @@ def save_file(request, file, type):
+            pdfrw.PdfWriter().write(filled_forms_path + dict_temp['Name'] + '_teem.pdf', template_pdf)
+
+            # Sends the email
+            send_mail(dict_temp['Mail'], "TEEM'21", dict_temp['Name'], filled_forms_path + dict_temp['Name'] + '_teem.pdf')
```
Some of these values are hardcoded (`TEEM'21`, `filled_forms_path + dict_temp['Name'] + '_teem.pdf'`) and may change depending on the subject of the PDF to fill. Can we make them configurable?

cbjuan commented on generator/views.py:
```diff
@@ -78,36 +82,56 @@ def save_file(request, file, type):
     temp = tempfile.TemporaryFile()
     zf = zipfile.ZipFile(temp, 'w', zipfile.ZIP_DEFLATED)
-    files2zip_list = glob.glob(filled_forms_path + 'filled_form_*.pdf')
+    files2zip_list = glob.glob(filled_forms_path + '*_teem.pdf')
```
Again. Can you make the hardcoded values configurable?

cbjuan reviewed (commented):
Thank you for the PR! I'm glad you find this repo so useful to spend time updating everything and adding new features.

I added some comments to the PR to help you polish some things. In general terms, I'm OK with adding the new feature and with the cleanup/update you included.

One question. Does this PR work with Python 3? Last time I used the project it worked with Python 2.

Thanks again. Lets keep in touch

JiahuiChen99 commented:
I'll modify the PR according to what you commented in the code review. For now, I hope that it works for TEEM 21 congress.
I'll make a those hardcoded values to be configurable 👍 

I had to update most of the libraries and dependencies because when I ran it in my machine, some of those were deprecated and were throwing warnings and issues. Specially, PDFJinja which caused some major problems. I replaced it with pdfwr.

This version works with Python 3.

> Thank you for the PR! I'm glad you find this repo so useful to spend time updating everything and adding new features.
> 
> I added some comments to the PR to help you polish some things. In general terms, I'm OK with adding the new feature and with the cleanup/update you included.
> 
> One question. Does this PR work with Python 3? Last time I used the project it worked with Python 2.
> 
> Thanks again. Lets keep in touch

JiahuiChen99 committed:
fix: unique pdf name

JiahuiChen99 committed:
Merge pull request #5 from JiahuiChen99/hotfix/unique_pdf

fix: unique pdf name

JiahuiChen99 committed:
Merge pull request #6 from JiahuiChen99/hotfix/block_form

Hotfix/block form

JiahuiChen99 committed:
Merge pull request #7 from JiahuiChen99/hotfix/unique_pdf

Hotfix/unique pdf

JiahuiChen99 committed:
Merge pull request #8 from JiahuiChen99/master

Merge pull request #5 from JiahuiChen99/hotfix/unique_pdf

JiahuiChen99 committed:
style: remove unused code

JiahuiChen99 committed:
feat: html conference name form field

JiahuiChen99 committed:
refactor: conditional fields errors rendering with CSS style

JiahuiChen99 committed:
fix: fields styles conditions

JiahuiChen99 committed:
feat: conference form field

JiahuiChen99 committed:
feat: embed variable conference name into the generated certificate

JiahuiChen99 committed:
feat: embed variable conference name into the mail body

JiahuiChen99 committed:
chore: remove unused dependencies

JiahuiChen99 committed:
feat: filter unsupported conference name characters out

JiahuiChen99 committed:
fix: remove original template and data files

JiahuiChen99 committed:
refactor: generate compressed certificates folder

JiahuiChen99 committed:
feat: add automailer form property

JiahuiChen99 committed:
feat: get automailer form property from the request

JiahuiChen99 committed:
feat: automate mailing on automailer form property

JiahuiChen99 committed:
feat: add automailer property in the form view

JiahuiChen99 committed:
feat: automailer tooltip advise information

JiahuiChen99 committed:
chore: removed unused dependencies

JiahuiChen99 committed:
Merge pull request #9 from JiahuiChen99/fix/juan_requests

Fix/juan requests

JiahuiChen99 commented:
Hi @cbjuan !
Sorry for the late modifications, I've been busy finishing my bachelor's degree 😅 

I added the suggested modifications and also improved the application. Here you have a rundown of the changes. 

### This PR includes the following features and fixes

- Conference name form property
- Auto-mailing enabling
- Improved form error highlighting
- Mailing feature
- Unused, deprecated dependencies removal

On application start-up, it loads the mailing credentials from os environment variables defined in the **.env** configuration file. The file must be located in the **mailing** folder.

```bash
django_pdf_form_filler
├───...
├───mailing
│   └───.env
├───...
```

Two variables must be set in the .env config file:

- SENDER_EMAIL: Gmail account
- SENDER_PSW: Gmail password

cbjuan reviewed (approved):
Thanks for your PR, @JiahuiChen99. Changes look good to me

JiahuiChen99 commented:
Great, thank you @cbjuan !
